### PR TITLE
Direct contract function call

### PR DESCRIPTION
**Pull request by janezpodhostnik**
ref: https://github.com/onflow/cadence/issues/888

depends on this cadence PR: https://github.com/onflow/cadence/pull/878

Changes deducting transaction fees from a meta transaction to a direct contract function call. This bypasses a lot of boilerplate that is run during meta-transaction invocation.

This does not directly implement  https://github.com/onflow/cadence/issues/859 but is the first step to it. To implement the final solution, the hostEnv reference needs to stay the same (cross transactions) in the FVM execution, or should be settable in the interpreter. Both of those solutions require a lot more work.

This PR has two side effects:
- fee deduction should be faster
- events (from fee deduction) should be properly triggered

### CODE REVIEW

File: fvm/transactionContractFunctionInvocator.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+		)
+		defer span.Finish()
+	}
+
+	env := newEnvironment(*ctx, vm, sth, programs)
+	predeclaredValues := valueDeclarations(ctx, env)
+
+	env.setTransaction(proc.Transaction, proc.TxIndex)
+	env.setTraceSpan(span)
+	location := common.StringLocation("ContractFunctionInvocation")
```
Comment (janezpodhostnik) [inline fvm/transactionContractFunctionInvocator.go:59]: @turbolent What would be an appropriate location here?

File: fvm/fvm.go [hunk 2]
```diff
@@ -93,3 +97,29 @@ func (vm *VirtualMachine) invokeMetaTransaction(ctx Context, tx *TransactionProc
...
+	sth *state.StateHolder,
+	programs *programs.Programs,
+) (fvmErr errors.Error, processErr error) {
+
+	invocator := NewTransactionContractFunctionInvocator(
+		contractLocation,
+		functionName,
+		arguments,
+		argumentTypes,
+		zerolog.Nop(),
```
Comment (ramtinms) [inline fvm/fvm.go:120]: any reason not pass the real logger here?

File: fvm/transactionContractFunctionInvocator.go (hunk 1 above)
Comment (ramtinms) [reply fvm/transactionContractFunctionInvocator.go:60]: ideally this should be constant somewhere

File: fvm/transactionFeeDeductor.go [hunk 3]
```diff
@@ -37,23 +49,27 @@ func (d *TransactionFeeDeductor) deductFees(
...
+		parentSpan,
 		sth,
 		programs,
 	)
-	if txErr == nil {
-		proc.Events = append(proc.Events, feeTx.Events...)
-		proc.ServiceEvents = append(proc.ServiceEvents, feeTx.ServiceEvents...)
-		proc.Logs = append(proc.Logs, feeTx.Logs...)
-		proc.GasUsed = proc.GasUsed + feeTx.GasUsed
-	}
```
Comment (ramtinms) [inline fvm/transactionFeeDeductor.go:56]: We need a substitution for this behaviour

File: fvm/transactionContractFunctionInvocator.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+		span.LogFields(
+			traceLog.String("transaction.ContractFunctionCall", fmt.Sprintf("%s.%s", i.contractLocation.String(), i.functionName)),
+		)
+		defer span.Finish()
+	}
+
+	env := newEnvironment(*ctx, vm, sth, programs)
+	predeclaredValues := valueDeclarations(ctx, env)
+
+	if span != nil {
```
Comment (ramtinms) [inline fvm/transactionContractFunctionInvocator.go:57]: not sure if we need a condition, if is empty value should work fine.

File: fvm/transactionContractFunctionInvocator.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		i.argumentTypes,
+		runtime.Context{
+			Interface:         env,
+			Location:          location,
+			PredeclaredValues: predeclaredValues,
+		},
+	)
+
+	if err != nil {
+		i.logger.Info().
```
Comment (ramtinms) [inline fvm/transactionContractFunctionInvocator.go:75]: maybe warn level?

**Review by ramtinms — CHANGES_REQUESTED** [review]
I think we need to capture events.
I think we also need further tests for this functionality to check if is actually balance of account has changed after it and a test to make sure this method is not callable by user

File: fvm/transactionContractFunctionInvocator.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+		arguments:        arguments,
+		argumentTypes:    argumentTypes,
+		logger:           logger,
+	}
+}
+
+func (i *TransactionContractFunctionInvocator) Invoke(vm *VirtualMachine, ctx *Context, proc *TransactionProcedure, sth *state.StateHolder, programs *programs.Programs) (cadence.Value, error) {
+	var span opentracing.Span
+	if ctx.Tracer != nil && proc.TraceSpan != nil {
+		span = ctx.Tracer.StartSpanFromParent(proc.TraceSpan, trace.FVMExecuteTransaction)
```
Comment (ramtinms) [inline fvm/transactionContractFunctionInvocator.go]: I think it shouldn't be `FVMExecuteTransaction` should be some other constant showing invoke meta transaction is happening

**Comment by janezpodhostnik** [comment]
Thanks for the review @ramtinms. I see what you mean with losing the events if I create a new hostEnv and don't merge it properly.

I think transaction fee deduction might need to go inside of TransactionIvocator and accept  the host env as a parameter. Let me change this up and show you what I mean.

**Comment by janezpodhostnik** [comment]
This targets branch https://github.com/onflow/flow-go/pull/710 to get the latest cadence updates (this PR is required: https://github.com/onflow/cadence/pull/878 )

File: fvm/transactionInvocator.go [hunk 7]
```diff
@@ -196,7 +204,44 @@ func (i *TransactionInvocator) Process(
...
+		[]sema.Type{
+			sema.AuthAccountType,
+		},
+		zerolog.Nop(),
+	)
+	_, err := invocator.Invoke(env, proc)
+
+	if err != nil {
+		// TODO: Fee value is currently a constant. this should be changed when it is not
+		return errors.NewTransactionFeeDeductionFailedError(proc.Transaction.Payer, DefaultTransactionFees.ToGoValue().(uint64), err)
```
Comment (ramtinms) [inline fvm/transactionInvocator.go:230]:
```
DefaultTransactionFees.ToGoValue().(uint64)
```

this is prone to accidental changes, I think we should wrap the type casting with ok argument and gracefully return with out deducting fees.

File: fvm/transactionInvocator.go [hunk 8]
```diff
@@ -158,9 +163,12 @@ func (i *TransactionInvocator) Process(
...
 
-	// check the storage limits
-	if ctx.LimitAccountStorage && txError == nil {
-		txError = NewTransactionStorageLimiter().Process(vm, ctx, proc, sth, programs)
+	if txError == nil {
+		txError = i.checkAccountStorageLimit(vm, ctx, proc, sth, programs)
+	}
+
+	if txError == nil {
+		txError = i.deductTransactionFees(env, proc)
```
Comment (ramtinms) [inline fvm/transactionInvocator.go:171]: I guess we need to change this logic, when it fails we still have to charge fees right?

File: fvm/fvm_test.go [hunk 9]
```diff
@@ -1893,3 +1893,208 @@ func TestSigningWithTags(t *testing.T) {
...
+					txBody,
+					accounts[0],
+					privateKeys[0],
+				)
+				require.NoError(t, err)
+
+				tx = fvm.Transaction(txBody, 1)
+
+				err = vm.Run(ctx, tx, view, programs)
+				require.NoError(t, err)
```
Comment (ramtinms) [inline fvm/fvm_test.go:2087]: could you please add extra checks here to check if token transfer events are captured when fees are deducted.

File: fvm/fvm_test.go [hunk 10]
```diff
@@ -1893,3 +1893,208 @@ func TestSigningWithTags(t *testing.T) {
 	}
 
 }
+
+func TestTransactionFeeDeduction(t *testing.T) {
```
Comment (ramtinms) [inline fvm/fvm_test.go:1897]: nice test 👏

**Review by ramtinms — APPROVED** [review]

File: fvm/fvm_test.go (hunk 9 above)
Comment (janezpodhostnik) [reply fvm/fvm_test.go:2087]: this test case
```
name:          "Transaction fee deduction emits events",
```

already does that. Unless you had something more specific in mind.

File: fvm/transactionInvocator.go (hunk 8 above)
Comment (janezpodhostnik) [reply fvm/transactionInvocator.go:171]: I was going to leave that change for another PR. I don't want to include that change in this spork.

File: fvm/transactionContractFunctionInvocator.go [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+		functionName:     functionName,
+		arguments:        arguments,
+		argumentTypes:    argumentTypes,
+		logger:           logger,
+	}
+}
+
+func (i *TransactionContractFunctionInvocator) Invoke(vm *VirtualMachine, ctx *Context, proc *TransactionProcedure, sth *state.StateHolder, programs *programs.Programs) (cadence.Value, error) {
+	var span opentracing.Span
+	if ctx.Tracer != nil && proc.TraceSpan != nil {
```
[the remark this answers is not in the document]
Comment (ramtinms) [reply fvm/transactionContractFunctionInvocator.go]: looks good to me now.

File: fvm/fvm_test.go (hunk 9 above)
Comment (ramtinms) [reply fvm/fvm_test.go:2085]: ah I see, somehow I missed it. sorry

File: fvm/fvm_test.go [hunk 12]
```diff
@@ -1891,3 +1891,208 @@ func TestSigningWithTags(t *testing.T) {
...
```
Comment (turbolent) [inline fvm/fvm_test.go:1895]: Great tests!

File: fvm/transactionContractFunctionInvocator.go [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+	predeclaredValues := valueDeclarations(&env.ctx, env)
+
+	value, err := env.vm.Runtime.InvokeContractFunction(
+		i.contractLocation,
+		i.functionName,
+		i.arguments,
+		i.argumentTypes,
+		runtime.Context{
+			Interface:         env,
+			Location:          contractFunctionInvocationLocation,
```
Comment (turbolent) [inline fvm/transactionContractFunctionInvocator.go:63]: Is it possible to pass nothing, or if not, `i.contractLocation`?

File: fvm/transactionContractFunctionInvocator.go [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+		logger:           logger,
+	}
+}
+
+func (i *TransactionContractFunctionInvocator) Invoke(env *hostEnv, proc *TransactionProcedure) (cadence.Value, error) {
+	var span opentracing.Span
+	if env.ctx.Tracer != nil && proc.TraceSpan != nil {
+		span = env.ctx.Tracer.StartSpanFromParent(proc.TraceSpan, trace.FVMInvokeContractFunction)
+		span.LogFields(
+			traceLog.String("transaction.ContractFunctionCall", fmt.Sprintf("%s.%s", i.contractLocation.String(), i.functionName)),
```
Comment (turbolent) [inline fvm/transactionContractFunctionInvocator.go:49]: Minor: Could we prepare this once so we don't have to prepare it on each trace?

**Review by turbolent — APPROVED** [review]
Nice, looks good!

File: fvm/transactionContractFunctionInvocator.go (hunk 13 above)
Comment (janezpodhostnik) [reply fvm/transactionContractFunctionInvocator.go]: Didn't know it was optional. Fixed to pass nothing

File: fvm/transactionContractFunctionInvocator.go (hunk 14 above)
Comment (janezpodhostnik) [reply fvm/transactionContractFunctionInvocator.go]: I created a field on the TransactionContractFunctionInvocator to hold the LogFields

File: fvm/transactionContractFunctionInvocator.go (hunk 13 above)
Comment (turbolent) [reply fvm/transactionContractFunctionInvocator.go]: I just made it optional yesterday, as @ramtinms was asking me what to pass in his case where he's calling `ReadStored`

**Review by turbolent — APPROVED** [review]

**Pull request merged**